Add device tracker for TP-Link EAP

chemicalstorm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Create new device tracker for TP-Link EAP access point.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/15146

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chemicalstorm commented:
Hum, just saw https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md which seems to mean that new integrations should use config flows.
It was not mentioned in the dev documentation, I guess I could update that part too then :)

audricschiltknecht committed:
Scaffold TP-Link EAP

MartinHjelmare commented on homeassistant/components/tp_link_eap/device_tracker.py:
```diff
@@ -0,0 +1,114 @@
+        """
+        _LOGGER.debug("Scanning...")
+
+        session = pxssh.pxssh(echo=False, encoding="utf-8")
```
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

chemicalstorm replied:
Sorry, I basically copy&pasted an existing device tracker and did not realize that it was not allowed anymore for new component.

MartinHjelmare reviewed (changes requested):
Besides a config flow the device specific interface needs to move to a 3rd party library.

audricschiltknecht committed:
Implement TP-Link EAP device tracker.

chemicalstorm commented:
Rewriten integration using 3rd party dep in pypi and config flow.

chemicalstorm commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+    coordinator.tracked_entity_macs = tracked
+
+    if not known_entities:
+        # If no entities exist, need to bootstrap process
```
I'm not 100% sure why I need to do that here.
If I remove the test protect, then it looks like `coordinator.async_refresh()` is called twice.
If I remove the call to `coordinator.async_refresh()` altogether, then when there are no already defined entities (eg. after setting up the integration), then the coordinator appears to not be updated...

MartinHjelmare replied:
The coordinator will only refresh automatically when there are entities that have connected callbacks to the coordinator. So if we need to update without entities, to be able to find entities, we may need to make sure there's at least one callback. We could eg make a callback for the entity addition instead of building that into the coordinator update function.

Here's the interface to add a refresh callback to the coordinator:
https://github.com/home-assistant/core/blob/1bb535aa6784fcb1af69749ec8eea2a1207456b9/homeassistant/helpers/update_coordinator.py#L77

chemicalstorm commented on homeassistant/components/tplink_eap/__init__.py:
```diff
@@ -0,0 +1,57 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up TP-Link EAP from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
```
Is it expected that we initialize the data for the domain as currently done? It was not present in the scaffolded files and without it, `hass.data[DOMAIN]` raises a `KeyError` when trying to store the API object.

frenck replied:
This is fine 👍

frenck commented on homeassistant/components/tplink_eap/strings.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "title": "TP-Link EAP",
```
Does the title need to be translated? It is a brand/product name, I guess that cannot be translated into languages right?

chemicalstorm replied:
Correct, I removed it from the file

frenck commented on homeassistant/components/tplink_eap/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/tplink_eap",
+  "requirements": ["pytleap==1.0.1"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
Remove empty defaults

```suggestion
```

frenck commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+"""Tracker for Wifi clients connected to a TP-Link EAP."""
```
If you are not going to provide tests for this file, it should be put into the `.coveragerc` file.

frenck commented on homeassistant/components/tplink_eap/__init__.py:
```diff
@@ -0,0 +1,57 @@
+
+async def async_setup(hass: HomeAssistant, config):
+    """Set up the TP-Link EAP component.
+
+    Configuration through YAML is not supported at this time.
```
And will never be, as we don't allow that anymore.

```suggestion
```

frenck reviewed (commented):
A couple of initial comments from me. Please note, that we require the configuration flow to have 100% test coverage. Currently, a couple of lines are not covered yet, so you might need to add a couple of tests for those.

audricschiltknecht committed:
Address PR comments

- Exclude device_tracker from coverage.
- Remove empty default value/useless translation strings.
- 100% coverage on config_flow.

chemicalstorm commented:
I've addressed the initial comments. For now I've excluded the `device_tracker` from coverage, hoping I'll find some time to look into implementing tests for it.

MartinHjelmare commented on homeassistant/components/tplink_eap/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+            try:
+                info = await validate_input(self.hass, user_input)
+                return self.async_create_entry(**info)
```
Please only wrap the line that can raise in the try... except block. Use an `else:` block if needed.

MartinHjelmare commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+    tracked: Set[str] = set()
+    registry = await entity_registry.async_get_registry(hass)
+    known_entities: List[Entity] = []
+    for entity in registry.entities.values():
```
```suggestion
    for entity_entry in registry.entities.values():
```

MartinHjelmare commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+        return self.mac
+
+    @property
+    def state_attributes(self):
```
Implement the standard `mac_address` property instead.

https://developers.home-assistant.io/docs/core/entity/device-tracker#properties

MartinHjelmare commented on homeassistant/components/tplink_eap/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
```
We're using url in the config flow, not host.

MartinHjelmare commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+            update_method=self.async_update,
+        )
+
+    async def async_update(self):
```
When subclassing the coordinator we should override `_async_update_data`:
https://github.com/home-assistant/core/blob/1bb535aa6784fcb1af69749ec8eea2a1207456b9/homeassistant/helpers/update_coordinator.py#L135

MartinHjelmare commented on homeassistant/components/tplink_eap/device_tracker.py:
```diff
@@ -0,0 +1,135 @@
+        if new_tracked:
+            self.async_add_entities(new_tracked, True)
+
+        self.connected_clients = {c.mac_address: c for c in connected_clients}
```
Return this data from the update method instead.

chemicalstorm commented:
Unfortunately I've not found the time to fully address all the comments, and I've gradually moved most of my device tracking to using the app instead. I'll be closing this PR then.
